### PR TITLE
User Profile and User Management

### PR DESCRIPTION
**Pull request by Abdullah-Anaz**

### CODE REVIEW

File: public/i18n/en/translation.json [hunk 1]
```diff
@@ -44,5 +44,6 @@
   "Time Out": "Time Out",
   "Copyright Elzian Agro. All Rights Reserved": "Copyright {{val}} Elzian Agro. All Rights Reserved",
   "Page Not Found": "Page Not Found",
-  "Go Home": "Go Home"
+  "Go Home": "Go Home",
+  "Email address not verified": "Email address not verified"
```
Comment (Abdullah-Anaz) [inline public/i18n/en/translation.json]: Add translations for profile and edit user page

File: src/App.js [hunk 2]
```diff
@@ -1,27 +1,34 @@
...
   return (
     <Router>
       <Routes>
-        <Route path='/' element={<LoginPage />} />
-        <Route path='/dashboard' element={<Dashboard />} />
+        <Route element={<PrivateRoutes />}>
+          <Route path='/dashboard' element={<Dashboard />} />
+
+          {/* Pages */}
+          <Route path='/users' element={<Dashboard page={<ManageUsers />} />} />
```
Comment (Abdullah-Anaz) [inline src/App.js]: Only render if userType == admin

File: src/App.js [hunk 3]
```diff
@@ -1,27 +1,34 @@
...
+          <Route path='/profile' element={<Dashboard page={<UserProfilePage />} />} />
 
-        {/* Pages */}
-        <Route path='/customer' element={<Dashboard />} />
-        <Route path='/orders' element={<Dashboard />} />
-        <Route path='/analytics' element={<Dashboard />} />
-        <Route path='/settings' element={<Dashboard />} />
-        <Route path='/addProducts' element={<Dashboard />} />
+          {/* Charts */}
+          <Route path='/pie' element={<Dashboard />} />
```
Comment (Abdullah-Anaz) [inline src/App.js]: Remove unneccessary routes
user needs to have dashboard (device analytics), weather, device, profile
admin needs to have dashboard (user analytics, device analytics), weather, device management, user management, profile

File: src/components/common/sidebar/index.jsx [hunk 4]
```diff
@@ -17,6 +17,16 @@ const Sidebar = () => {
 
   const currentYear = new Date().getFullYear();
 
+  const logout = () => {
+    try {
+      localStorage.removeItem('jwtAccessToken');
+      localStorage.removeItem('jwtRefreshToken');
+    } catch (error) {
+      console.error('Error during logout:', error);
```
Comment (Abdullah-Anaz) [inline src/components/common/sidebar/index.jsx]: remove console.error and handle error accordingly

File: src/components/common/sidebar/sidebarConstants.js [hunk 5]
```diff
@@ -28,9 +34,9 @@ const sidebarLinks = [
...
   },
   {
-    to: '/addProducts',
-    text: 'Add Products',
-    icon: PlusIcon,
+    to: '/profile',
+    text: 'My Profile',
+    icon: UserCircleIcon,
   },
 ];
```
Comment (Abdullah-Anaz) [inline src/components/common/sidebar/sidebarConstants.js:12]: check the userType

File: src/pages/dashboard/components/base/Button.jsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        {text}
+      </span>
+      <span className='absolute right-0 h-full w-12 sm:w-12 rounded-lg bg-green-500 flex items-center justify-center transform group-hover:translate-x-0 group-hover:w-full transition-all duration-300'>
+        {Icon && <Icon className='h-5 w-5 text-white mr-2' />}
+      </span>
+    </button>
+  );
+};
+
+const ToggleButton = ({ value, onChange }) => {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/base/Button.jsx:32]: make it a bit smaller

File: src/pages/dashboard/components/base/TextBox.jsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { EyeIcon, EyeSlashIcon } from '@heroicons/react/24/outline';
+
+function TextBox({ label = null, type = 'text', placeholder = '', Icon = null, value, setValue }) {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/base/TextBox.jsx]: make the element smaller
reduce padding and increase margin

File: src/pages/dashboard/components/common/alert-box/index.jsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
+import { PrimaryButton } from '../../base/Button';
+import PropTypes from 'prop-types';
+
+const AlertBox = ({ visible, message, onClose }) => {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/alert-box/index.jsx]: dark theme => bg-black
light theme => bg-white

File: src/pages/dashboard/components/common/confirm-box/index.jsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { PrimaryButton } from '../../base/Button';
+import PropTypes from 'prop-types';
+
+const ConformBox = ({ visible, message, onClose }) => {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/confirm-box/index.jsx]: dark theme => bg-black
light theme => bg-white

File: src/pages/dashboard/components/common/form/index.jsx [hunk 10]
```diff
@@ -0,0 +1,285 @@
+import React, { useState, useEffect } from 'react';
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/form/index.jsx]: rename folder: form -> user-form

File: src/pages/dashboard/components/common/form/index.jsx [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    if (!user) {
+      if (!password.trim()) {
+        newErrors.password = '*Password is required';
+      }
+    }
+
+    if (!orgName.trim()) {
+      newErrors.orgName = '*Organization name is required';
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/form/index.jsx]: all the errors and hard coded values should be imported from constant files
use .map() for the validation function

File: src/pages/dashboard/components/common/form/index.jsx [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+
+    const accessToken = localStorage.getItem('jwtAccessToken');
+
+    const requestData = {
+      firstName,
+      lastName,
+      NIC: nic,
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/form/index.jsx]: RENAME: NIC -> nic

File: src/pages/dashboard/components/common/form/index.jsx [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    if (validateForm()) {
+      const apiEndpoint = user
+        ? `${process.env.REACT_APP_BASE_URL}/user/${user._id}`
+        : `${process.env.REACT_APP_BASE_URL}/user`;
+
+      const method = user ? 'put' : 'post';
+
+      axios[method](apiEndpoint, requestData, {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/form/index.jsx]: custom hook for axios

File: src/pages/dashboard/components/common/loader/index.jsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+function Loader() {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/loader/index.jsx:3]: change style
decrease size

File: src/pages/dashboard/components/common/notification/index.jsx [hunk 15]
```diff
@@ -1,25 +1,146 @@
...
+  const allNotifications = useSelector(selectAllNotifications);
+  const token = localStorage.getItem('jwtAccessToken');
 
   const closeNotification = () => {
     dispatch(setNotificationOpen(false));
   };
 
+  const readAllNotifications = async () => {
+    try {
+      await axios.put(
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/notification/index.jsx]: custom hook

File: src/pages/dashboard/components/common/notification/index.jsx [hunk 16]
```diff
@@ -1,25 +1,146 @@
...
+      dispatch(setNotificationsCount(null));
+    } catch (error) {
+      console.error('Error marking all notifications as read', error);
+    }
+  };
+
+  const deleteNotification = async (index) => {
+    // Delete from the database
+    try {
+      await axios.delete(`${process.env.REACT_APP_BASE_URL}/notification/`, {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/notification/index.jsx]: use custom hook

File: src/pages/dashboard/components/common/notification/index.jsx [hunk 17]
```diff
@@ -1,25 +1,146 @@
...
+
+    // Change the flage to true on local list.
+    // Deep cloning because the spread operator's shallow copy may not work as expected with nested objects.
+    const updatedNotifications = JSON.parse(JSON.stringify(allNotifications));
+    updatedNotifications[index].read = true;
+    dispatch(setAllNotifications(updatedNotifications));
+
+    //Change the flage to true on database
+    try {
+      await axios.put(
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/components/common/notification/index.jsx]: use custom hook

File: src/pages/dashboard/manage-users/index.jsx [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+          }}
+        />
+      ),
+    },
+  ];
+
+  const getUsers = useCallback(async () => {
+    try {
+      const { jwtAccessToken: accessToken } = localStorage;
+      const response = await axios.get(`${process.env.REACT_APP_BASE_URL}/user`, {
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/manage-users/index.jsx]: use custom hook

File: src/pages/dashboard/manage-users/index.jsx [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+    if (result) {
+      handleDelete();
+    }
+    setIsConfirmVisible(false);
+  };
+
+  const handleDelete = async () => {
+    try {
+      const { jwtAccessToken: accessToken } = localStorage;
+      await axios.put(
```
Comment (Abdullah-Anaz) [inline src/pages/dashboard/manage-users/index.jsx]: use custom hook

File: src/pages/dashboard/utils/generateRandomPassword.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
+export const generateRandomPassword = (length = 12) => {
+  const lowercaseLetters = 'abcdefghijklmnopqrstuvwxyz';
```
Comment (pankajan05) [inline src/pages/dashboard/utils/generateRandomPassword.js]: bring this from the constant file

File: src/pages/utils/GetNotifications.js [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+import { setAllNotifications, setNotificationsCount } from 'pages/dashboard/slice/notificationSlice';
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
+
+const GetNotifications = () => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    const getNotifications = async () => {
+      const token = localStorage.getItem('jwtAccessToken');
```
Comment (pankajan05) [inline src/pages/utils/GetNotifications.js]: No need to create a variable to store this. can access the token directly

File: src/pages/utils/GetNotifications.js [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+        const datas = response.data.result;
+
+        // Count the number of objects where readFlag is true
+        const readNotificationsCount = datas.filter((data) => !data.readFlag).length;
+        dispatch(setNotificationsCount(readNotificationsCount));
+
+        const notifications = datas.map((data) => {
+          // get the Date from the timestampString
+          const timestampString = data.dateTime;
```
Comment (pankajan05) [inline src/pages/utils/GetNotifications.js]: This functionality can move to backend.

File: src/pages/utils/GetUserData.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+        const token = localStorage.getItem('jwtAccessToken');
+
+        const response = await axios.get(`${process.env.REACT_APP_BASE_URL}/user/profile`, {
+          headers: {
+            Authorization: `Bearer ${token}`,
+          },
+        });
+
+        dispatch(setUserData(response.data));
+        console.log(response.data);
```
Comment (pankajan05) [inline src/pages/utils/GetUserData.js]: Remove the console.log

File: src/pages/utils/GetUserData.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+      } catch (error) {
+        console.error('Error fetching user data:', error);
```
Comment (pankajan05) [inline src/pages/utils/GetUserData.js]: Not good to use console.error

File: src/pages/dashboard/user-profile-page/index.js [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+            Authorization: `Bearer ${token}`,
+          },
+        }
+      );
+
+      // Close the address edit mode and reset the local state for newAddress
+      setAddressEditMode(false);
+      setNewAddress('');
+    } catch (error) {
+      console.error('Error updating address:', error);
```
Comment (pankajan05) [inline src/pages/dashboard/user-profile-page/index.js]: Error does not handle correctly.

File: src/pages/dashboard/user-profile-page/index.js [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+
+        // Make Axios PUT request to the specified endpoint
+        await axios.put(`${process.env.REACT_APP_BASE_URL}/user/profile/image`, formData, {
+          headers: {
+            'Content-Type': 'multipart/form-data',
+            Authorization: `Bearer ${token}`,
+          },
+        });
+      } catch (error) {
+        console.error('Error uploading profile picture:', error);
```
Comment (pankajan05) [inline src/pages/dashboard/user-profile-page/index.js]: Error does not handle correctly.

**Comment by pankajan05** [comment]
Team review on creating the custom hook for the axios. In that way you don't need to access the token in multiple files, and you can do the error handling simply.

File: public/i18n/en/translation.json (hunk 1 above)
Comment (tkdeshan) [reply public/i18n/en/translation.json]: for user edit page added

**Review by pankajan05 — APPROVED** [review]
Checked and Suggested the improvements!

**Pull request merged**